Massaging of Intro and early parts of the I-D in prep for WGLC (#40)
cose-wg/draft-ietf-cose-hash-envelope: draft-ietf-cose-hash-envelope.md

# Introduction

COSE defined detached payloads in Section 2 of {{-COSE}}, using `nil` as the payload.

> In order to verify a signature over a detached payload, the signature checker requires access to the payload content.

Reviewer: ```suggestion
In order to verify a signature over a cose-sign1, the signature checker requires access to the payload content.
```

Hashes are already used on a regular basis as identifiers for payload data, such as documents or software components.
As hashes typically are smaller than the payload data they represent, they are simpler to transport.
Additional hints in the protected header ensure cryptographic agility for the hashing and signing algorithms.

> When hashes are used as identifiers, hints also improve discoverability of original payload content the hashes represent.

Reviewer: ```suggestion
Hashes and other identifiers are commonly used as hints to discover and distinguish resources.
Using a hash as an identifier for a resource has the advantage of enabling integrity checking.
```

> In some applications, such as remote signing procedures, conveyance of hashes instead original payload content reduce the conveyed amount of data drastically.

Reviewer: ```suggestion
In some applications, such as remote signing procedures, conveyance of hashes instead original payload content reduce transmission time and costs.
```

# Header Parameters

> This document specifies the following new header parameters commonly used to identify resources that are hashed:

Reviewer: ```suggestion
This document specifies the following new header parameters commonly used alongside hashes to identify resources:
```

# Header Parameters

To represent a hash of a payload, the following headers are defined:
This document specifies the following new header parameters commonly used alongside hashes to identify resources:

TBD_1:
  : the hash algorithm used to produce the payload.

TBD_2:
  : the content type of the bytes that were hashed to produce the payload.
  : the content type of the bytes that were hashed (preimage) to produce the payload.

TBD_3:
  : an identifier enabling a verifier to retrieve the bytes which were hashed to produce the payload.
  : an identifier enabling retrieval of the original resource (preimage) identified by the payload.

# Hash Envelope CDDL

~~~ cddl
Hash_Envelope_Protected_Header = {
    ? &(alg: 1) => int,

>     ? &(typ: 16) => uint / tstr

Reviewer 2: why is typ 16 removed?

Reviewer: Its not needed in this document.
There are a bunch of other header parameters that are also not needed here, which can be found at:

https://www.iana.org/assignments/cose/cose.xhtml#header-parameters (kid, crit, etc)...

Hash_Envelope = #6.18(Hash_Envelope_as_COSE_Sign1)
~~~

- Label `1` (alg) Cryptographic algorithm to use
- Label `TBD_1` (payload hash alg) MUST be present in the protected header and MUST NOT be present in the unprotected header.
- Label `TBD_2` (content type of the preimage of the payload) MAY be present in the protected header or unprotected header.

> - Label `TBD_3` (payload_location) MAY be added to the protected header and MUST NOT be presented in the unprotected header.

Reviewer: Not sure the payload location should be prohibited from the unprotected header.

I feel like we talked about this already... 

I think given these area all optional except for TBD_1, when present they should all be in the protected header, or they should all be allowed to be in either protected or unprotected, but not both.

Reviewer 2: If a hint is unprotected, it also means someone could add it, redirecting the user to a hacked location. Protecting the location says the issuer states it is/was at that location, although the verifier may not have access to it. 

As the unprotected header is stripped, it would render it minimally useful, and don't think we should suggest it could be in the unprotected header

Reviewer: You can't stop people from putting stuff in the unprotected header.
The proper security response to seeing some untrusted link, is not to click it.

By the same argument you give for location, I could trick a client that resolved the data to attempt to unzip it by setting the the content type to:

[application/vnd.cncf.helm.chart.content.v1.tar+gzip](https://www.iana.org/assignments/media-types/application/vnd.cncf.helm.chart.content.v1.tar+gzip)

Author: Is it better to not have a hint than an unprotected hint?

Author: I am not going to die on this hill: but...

something being optional does not imply that is not protection worthy, right? So there needs to be a better reason, I think.

Reviewer 3: If we define a header parameter, we can say what it is intended to be used for.
If we don't want it to be in the unprotected header, we can say so.
This should encourage implementations to not accept it there.
(This can be extended to a "MUST be ignored" in the unprotected header.)

Author: Is there any such apatite for such strong normative language? I don't think that we want to be restrictive in such a manner here. Any counter indications?

Reviewer 2: > given that content type and location are already optional, we admit that it is ok for them to be unprotected.
Being optional doesn't preclude being protected, so I admin nothing :)

I'd vote for leaving the text as-is, por favor

Reviewer 2: Resolving based on the above discussion. Please activate/block if you feel differently.
